TST: Fix assert raises in `sklearn/metrics/tests/`

sameshl opened the pull request:
replaced `assert_raises` and `assert_raises_regex` with `pytest.raises` context manager.

related to #14216

sameshl committed:
fix `assert_raises` in `test_regression.py`

related to #14216

sameshl committed:
fix `assert_raises` in `test_classification.py`

sameshl committed:
fix `assert_raises` in `test_common.py`

sameshl committed:
fix `assert_raises` and `assert_raises_regex` in `test_pairwise.py`

sameshl committed:
fix `assert_raises` and `assert_raises_regex` in `test_ranking.py`

sameshl committed:
fix `assert_raises` and `assert_raises_regex` in `test_score_objects.py`

glemaitre reviewed (changes requested):
Same as other PRs: remove `assert_raise_message`.

glemaitre commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -556,7 +555,7 @@ def test_not_symmetric_metric(name):
 
     # use context manager to supply custom error message
-    with assert_raises(AssertionError) as cm:
+    with pytest.raises(AssertionError) as cm:
```
This will not work as intended with pytest.

I think that we want something like this instead

```python
with pytest.raises(AssertionError):
    assert_array_equal(metric(y_true, y_pred), metric(y_pred, y_true))
    raise ValueError("%s seems to be symmetric:" %name)
```

```

sameshl replied:
@glemaitre Thanks for pointing that out! Will make the required change.

glemaitre commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1140,7 +1152,7 @@ def check_sample_weight_invariance(name, metric, y1, y2):
 
     # use context manager to supply custom error message
-    with assert_raises(AssertionError) as cm:
+    with pytest.raises(AssertionError) as cm:
```
This is the same issue than above. You need to raise a ValueError within the context manager.

glemaitre commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -90,15 +89,15 @@ def test_regression_metrics_at_limits():
-    assert_raises_regex(ValueError, "Mean Squared Logarithmic Error cannot be "
-                        "used when targets contain negative values.",
-                        mean_squared_log_error, [1., -2., 3.], [1., 2., 3.])
+    with pytest.raises(ValueError, match="Mean Squared Logarithmic Error "
```
You can move the message out of `raises`

glemaitre reviewed (dismissed):
Otherwise LGTM. We should open something to refactor these tests with parametrization.

sameshl committed:
fix use of `with pytest.raises` in test_common.py

sameshl committed:
move error message out of pytest.raises

sameshl committed:
fix `assert_raise_message` in test_common.py

sameshl committed:
fix `assert_raise_message` in test_pairwise.py

sameshl committed:
fix `assert_raise_message` in test_ranking.py

sameshl committed:
fix `assert_raise_message` in test_classification.py

sameshl committed:
fix linting error

sameshl commented:
@rth. I have made the required changes. You can have a look.

sameshl committed:
manually escape the string without the use of re.escape

sameshl commented:
@rth I have removed all except one instances of `re.escape` and removed them by manually escaping the string. The one I am not able to do is dynamically escape the string. Can you help me out on that?

sameshl commented:
@thomasjpfan Can you have a look?

rth reviewed (approved):
Thanks @sameshl !
